src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,397 @@
+# the prefix is free from non-prefix characters (the separator). Note that the prefix is expected to be at least two
+# characters in length to prevent overlap with separators and confusion with versions; the prefix separator is at most
+# three characters; and a negative lookback passes if there are no preceding characters.
+
+# The infix exists between parts of the version string. The most recent design resulted in use of a back reference to
+# ensure non-suffix version parts were separated by the same separator, e.g. 1.2.3 but not 1.2-3. However, one edge
+# case required this to be partially reverted, as it required 1.2-3 to be accepted, while another edge case where
+# additional zeros need to be added after a version to pad its length, e.g. 1.2 becomes 1.2.0.0, still requires it.
+
+# The infix accepts either:
+# - One to three alphabetic characters.
+# - One to three non-alphanumeric characters.
+# Note: The upper limit of three could be reduced to two based on current data.
+INFIX_3 = "([a-z]{1,3}|[^0-9a-z]{1,3})"
+INFIX_1 = f"(?P<sep>{INFIX_3})"  # A named capture group of INFIX_3.
+INFIX_2 = "(?P=sep)"  # A back reference to INFIX_1.
+
+# The suffix separator exists for much the same purpose as the prefix separator: splitting the suffix into the actual
+# suffix, and the characters that join it to the version.
+# It optionally accepts:
+# One to two non-alphanumeric characters that are followed by a non-numeric character (positive lookahead).
+SUFFIX_SEPARATOR = "(?P<suffix_sep>[^0-9a-z]{1,2}(?=[^0-9]))?"
+
+# The suffix optionally accepts:
+# A string that starts with an alphabetic character, and continues for one or more characters of any kind.
+SUFFIX = "(?P<suffix>[a-z].*)?"
+
+split_pattern = re.compile("[^0-9a-z]", flags=re.IGNORECASE)
+validation_pattern = re.compile("[0-9a-z]+", flags=re.IGNORECASE)
+alphabetic_only_pattern = re.compile("[a-z]+", flags=re.IGNORECASE)
+numeric_only_pattern = re.compile("[0-9]+")
+versioned_string = re.compile("[a-z]+[0-9]+", flags=re.IGNORECASE)  # e.g. RC1, M5, etc.
+
+
+def get_commit_from_version(git_obj: Git, purl: PackageURL) -> tuple[str, str]:
+    """Try to find the matching commit in a repository of a given version via tags.
+
+    The version of the passed PackageURL is used to match with the tags in the target repository.
+
+    Parameters
+    ----------
+    git_obj: Git
+        The repository.
+    purl: PackageURL | None
+        The PURL of the artifact.
+
+    Returns
+    -------
+    tuple[str, str]
+        The branch name and digest as a tuple.
+    """
+    if purl.version is None:
+        logger.debug("Missing version for artifact: %s", purl.name)
+        return "", ""
+    logger.debug("Searching for commit of artifact version using tags: %s@%s", purl.name, purl.version)
+
+    target_version_pattern, parts, has_non_numeric_suffix = _build_version_pattern(purl.version)
```

Comment:
Because these values are only relevant in the `_match_tags` function, I reckon it is better to move this call of `_build_version_pattern` into `_match_tags` instead of leaving it here.